### PR TITLE
Ensure zero token nodes are handled properly

### PR DESCRIPTION
**Pull request by sylwiaszunejko**
# Support zero-token nodes

Adds coverage for zero-token nodes (ScyllaDB nodes started with `join_ring: false`, which report
`tokens = null` in `system.peers`) and fixes a replica-placement bug found while writing it.

Expected semantics, aligned with the Rust driver: a zero-token node is an ordinary routable host —
discovered, kept in load balancing plans, usable as a coordinator and as a contact point — but it
never owns data, so it must never be returned as a replica. Processing one must not warn or error.

## Bug fix

`TokenMap.Build` registered every host in its per-datacenter host count and rack set, including
hosts that own no tokens. `NetworkTopologyStrategy` uses that rack set to decide whether racks are
still missing, so a zero-token node on its own rack inflated `Racks.Count`, kept `racksMissing`
true forever and caused real token owners to be skipped — leaving the replication factor
unsatisfied. The inflated host count also made `dcRf` too large, forcing a full ring scan per token.

The fix registers a datacenter only for hosts that actually own tokens. A datacenter made up
exclusively of zero-token nodes is therefore absent from the map and its replication factor is
treated as trivially satisfied. `HostLength`/`Racks` are consumed only by
`NetworkTopologyStrategy`, so the blast radius is limited to it.

Fixes: https://scylladb.atlassian.net/browse/DRIVER-191

### CODE REVIEW

File: ci/run-zero-token-manual-test.sh [hunk 1]
```diff
@@ -0,0 +1,100 @@
+#!/usr/bin/env bash
```
Comment (dkropachev) [inline ci/run-zero-token-manual-test.sh]: Why do we need that ?

File: tests/ZeroTokenManualTest/Program.cs [hunk 2]
```diff
@@ -0,0 +1,333 @@
+//
```
Comment (dkropachev) [inline tests/ZeroTokenManualTest/Program.cs]: Let's have these tests in regular integration suite

**Review by dkropachev — CHANGES_REQUESTED** [review]
Also make sure all the cases covered - including node being loaded with zero-token neds and then upgraded to the regular node.

File: src/Cassandra.IntegrationTests/Core/ZeroTokenNodeTests.cs [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+            for (var i = 0; i < ZeroTokenNodeTests.RoutingKeySamples; i++)
+            {
+                var rs = session.Execute($"SELECT k, v FROM {ZeroTokenNodeTests.TableName} WHERE k = {i}");
+                coordinators.Add(rs.Info.QueriedHost.Address.ToString());
+            }
+
+            CollectionAssert.Contains(
+                coordinators,
+                ZeroTokenAddress,
+                "a zero-token node owns no data but is still a usable coordinator");
```
[the remark this answers is not in the document]
Comment (sylwiaszunejko) [reply src/Cassandra.IntegrationTests/Core/ZeroTokenNodeTests.cs:184]: not true, they can serve as a coordinator only nodes

File: src/Cassandra.Tests/ZeroTokenNodeUnitTests.cs [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+
+            var policy = new RoundRobinPolicy();
+            policy.Initialize(clusterMock.Object);
+
+            // The zero-token host is still a valid coordinator...
+            Assert.AreEqual(HostDistance.Local, policy.Distance(zeroToken));
+
+            // ...and still appears in the query plan alongside the normal host.
+            var plan = policy.NewQueryPlan(null, new SimpleStatement()).Select(h => h.Host).ToList();
+            CollectionAssert.Contains(plan, zeroToken);
```
[the remark this answers is not in the document]
Comment (sylwiaszunejko) [reply src/Cassandra.Tests/ZeroTokenNodeUnitTests.cs:69]: not true, they can serve as a coordinator only nodes

File: Makefile [hunk 5]
```diff
@@ -57,18 +57,25 @@ test-unit: .use-development-snk
...
...ry_Parameters_CaseInsensitivity_NoOverload & FullyQualifiedName!~TokenAware_TransientReplication_NoHopsAndOnlyFullReplicas & FullyQualifiedName!~GetFunction_Should_Return_Most_Up_To_Date_Metadata_Via_Events & FullyQualifiedName!~LargeDataTests & FullyQualifiedName!~MetadataTests & FullyQualifiedName!~MultiThreadingTests & FullyQualifiedName!~PoolTests & FullyQualifiedName!~PrepareLongTests & FullyQualifiedName!~SpeculativeExecutionLongTests & FullyQualifiedName!~StressTests & FullyQualifiedName!~TransitionalAuthenticationTests & FullyQualifiedName!~ProxyAuthenticationTests & FullyQualifiedName!~CloudIntegrationTests & FullyQualifiedName!~CoreGraphTests & FullyQualifiedName!~GraphTests & FullyQualifiedName!~InsightsIntegrationTests & FullyQualifiedName!~DateRangeTests & FullyQualifiedName!~FoundBugTests & FullyQualifiedName!~GeometryTests & FullyQualifiedName!~LoadBalancingPolicyTests & FullyQualifiedName!~ConsistencyTests & FullyQualifiedName!~LoadBalancingPolicyTests & FullyQualifiedName!~ReconnectionPolicyTests & FullyQualifiedName!~RetryPolicyTests & FullyQualifiedName!~ZeroTokenNodeReplacementTests & FullyQualifiedName!~Should_KeepServingQueries_When_ZeroTokenNodeGoesDownAndUp)
```
Comment (dkropachev) [inline Makefile]: why these tests excluded from scylla integration test ?

File: Makefile [hunk 6]
```diff
@@ -57,18 +57,25 @@ test-unit: .use-development-snk
...
 .PHONY: test-integration-cassandra
 test-integration-cassandra: .use-development-snk .prepare-cassandra-ccm
 	CCM_DISTRIBUTION=cassandra dotnet test $(TEST_TARGET_OPTIONS) $(TEST_INTEGRATION_CSPROJ) $(TEST_INTEGRATION_OPTIONS)
 
+# Zero-token node integration tests only (runs full suite, including realclusterlong scenarios).
+# Override defaults with e.g.: make test-zero-token SCYLLA_VERSION=release:2026.1
+.PHONY: test-zero-token
+test-zero-token: .use-development-snk .prepare-scylla-ccm
+	dotnet build-server shutdown
+	CCM_DISTRIBUTION=scylla dotnet test $(TEST_TARGET_OPTIONS) $(TEST_INTEGRATION_CSPROJ) $(TEST_INTEGRATION_OPTIONS) --filter "FullyQualifiedName~ZeroTokenNode"
```
Comment (dkropachev) [inline Makefile]: There is no good reason for it, let's them be in the regular integration suit

**Review by dkropachev — CHANGES_REQUESTED** [review]

File: src/Cassandra.Tests/ZeroTokenNodeUnitTests.cs [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+            foreach (var tokenValue in new long[] { -100, 0, 5, 20, 500000 })
+            {
+                var replicas = tokenMap.GetReplicas("ks_nts", new M3PToken(tokenValue)).Select(r => r.Host).ToList();
+                CollectionAssert.AreEquivalent(new[] { withTokens1, withTokens2 }, replicas);
+                CollectionAssert.DoesNotContain(replicas, zeroToken);
+            }
+        }
+
+        [Test]
+        public void TokenMap_Should_IgnoreDatacenter_When_ItOnlyContainsZeroTokenHosts()
```
Comment (dkropachev) [inline src/Cassandra.Tests/ZeroTokenNodeUnitTests.cs:142]: Please add a unit test that rebuilds the token map after replacing the zero-token host with a token-owning host and verifies NTS starts placing replicas in this DC. The existing long integration test does not cover this transition in the unit suite.

File: Makefile (hunk 5 above)
Comment (sylwiaszunejko) [reply Makefile]: they take a little bit longer then the rest of the tests because of spinning up extra nodes etc, if that is not a problem I can include them

File: Makefile (hunk 6 above)
Comment (sylwiaszunejko) [reply Makefile]: ok

**Review by dkropachev — APPROVED** [review]
